verify/cert.go

```diff
@@ -49,40 +51,60 @@ func NewCertVerifier(cfg *Config, l log.Logger) (*CertVerifier, error) {
 	}, nil
 }
 
-// verifies on-chain batch ID for equivalence to certificate batch header fields
-func (cv *CertVerifier) VerifyBatch(
-	header *binding.IEigenDAServiceManagerBatchHeader, id uint32, recordHash [32]byte, confirmationNumber uint32,
+// verifyBatchConfirmedOnchain verifies that batchMetadata (typically part of a received cert)
+// matches the batch metadata hash stored on-chain
+func (cv *CertVerifier) verifyBatchConfirmedOnchain(
+	ctx context.Context, batchID uint32, batchMetadata *disperser.BatchMetadata,
 ) error {
-	blockNumber, err := cv.getConfDeepBlockNumber()
+	// 1. verify batch is actually onchain at the batchMetadata's state confirmedBlockNumber
+	// This assert is technically not necessary, but it's a good sanity check.
+	// It could technically happen that a confirmed batch's block gets reorged out,
+	// yet the tx is included in an earlier or later block, making this check fail...
+	confirmationBlockNumber := batchMetadata.GetConfirmationBlockNumber()
+	confirmationBlockNumberBigInt := big.NewInt(0).SetInt64(int64(confirmationBlockNumber))
+	_, err := cv.retrieveBatchMetadataHash(ctx, batchID, confirmationBlockNumberBigInt)
 	if err != nil {
-		return fmt.Errorf("failed to get context block: %w", err)
+		return fmt.Errorf("batch not found onchain at supposedly confirmed block %d: %w", confirmationBlockNumber, err)
```

Comment:
Is this true? If so I think we should get rid of this check but document exactly why we are getting rid of it. Would have saved me a lot of time if this had been documented in the initial code.

Reply:
Interesting. But I'm not even sure, because the case I'm describing is this:
0. (assumption: we wait for 5 block confirmation)
1. batcher confirms batch onchain at block 1000
2. eigenda-client wait for 5 blocks and returns a BlobStatus
3. proxy receives this one block later at block 1006, and the batch has actually been reorged and included in block 1001 instead.
4. In this case the check highlighted here would fail, even though the block is actually confirmed with depth 5 (the next assert below around line 70 would catch if that was not the case)

For this reason I think this check is actually wrong here, and I still think we should remove this check. Does this make sense?

Reply:
EigenDA batcher detects and updates confirmation number, if it detects reorg, https://github.com/Layr-Labs/eigenda/blob/master/disperser/batcher/finalizer.go#L198

Reply:
[55edfae](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/55edfaefb4297be0dbd31e6251db691b951560a4)